donkeycar/parts/datastore_v2.py

```diff
@@ -0,0 +1,367 @@
+            if not self.base_path.exists():
+                self.base_path.mkdir(parents=True, exist_ok=True)
+                print('Created a new datastore at %s' % (self.base_path.as_posix()))
+
+            self.seekeable = Seekable(self.manifest_path)
+
+        if not has_catalogs:
+            self._write_contents()
+            self._add_catalog()
+        else:
+            last_known_catalog = os.path.join(self.base_path, self.catalog_paths[-1]);
+            print('Using catalog %s' % (last_known_catalog))
+            self.current_catalog = Catalog(last_known_catalog, self.current_index)
+
+        # Register shutdown hook
+        atexit.register(self._exit_handler)
+
+    def write_record(self, record):
+        new_catalog = self.current_index > 0 and (self.current_index % self.max_len) == 0
+        if new_catalog:
+            self._add_catalog()
+
+        self.current_catalog.write_record(record)
+        self.current_index += 1
+        # Update metadata to keep track of the last index
+        self._update_catalog_metadata(update=True)
+
+    def delete_record(self, record_index):
+        # Does not actually delete the record, but marks it as deleted.
+        self.deleted_indexes.add(record_index)
+        self._update_catalog_metadata(update=True)
+
+    def _add_catalog(self):
+        current_length = len(self.catalog_paths)
+        catalog_name = 'catalog_%s.catalog' % (current_length)
+        catalog_path = os.path.join(self.base_path, catalog_name)
+        current_catalog = self.current_catalog
+        self.current_catalog = Catalog(catalog_path, start_index=self.current_index)
+        # Store relative paths
+        self.catalog_paths.append(catalog_name)
+        self._update_catalog_metadata(update=True)
+        if current_catalog:
+            current_catalog.close()
+
+    def _read_metadata(self, metadata=[]):
+        self.metadata = dict()
+        for (key, value) in metadata:
+            self.metadata[key] = value
+
+    def _read_contents(self):
+        self.seekeable.seek_line_start(1)
+        self.inputs = json.loads(self.seekeable.readline())
+        self.types = json.loads(self.seekeable.readline())
+        self.metadata = json.loads(self.seekeable.readline())
+        self.manifest_metadata = json.loads(self.seekeable.readline())
+        # Catalog metadata
+        catalog_metadata = json.loads(self.seekeable.readline())
+        self.catalog_paths = catalog_metadata['paths']
+        self.current_index = catalog_metadata['current_index']
+        self.max_len = catalog_metadata['max_len']
+        self.deleted_indexes = set(catalog_metadata['deleted_indexes'])
+
+    def _write_contents(self):
+        self.seekeable.truncate_until_end(0)
+        self.seekeable.writeline(json.dumps(self.inputs))
+        self.seekeable.writeline(json.dumps(self.types))
+        self.seekeable.writeline(json.dumps(self.metadata))
+        self.seekeable.writeline(json.dumps(self.manifest_metadata))
```

Comment:
These 4 lines seem to break json compatibility, as `self.inputs`, ..., `self.manifest_metadata` are all types like `list`, `dict`, etc. I think we need to pack these into a dict first and write this to the file like:
`data = dict(inputs=self.inputs, types=self.types, metadata=self.metadata, manifest_metadata=self.manifest_metadata)
self.seekable.writeline(json.dumps(data))
`
Otherwise the file `manifest.json` won't be a proper formatted json file.

Reply:
Ok, you are really stating that the file contains two lists, and 3 json objects (i.e. dicts). So probably we just don't name it .json? It probably also doesn't need to be a seekable.

Reply:
The `Manifest` is a `Seekable` because we want to make sure updates are very quick. Truncating the file is a much faster operation than doing a full flush anytime we need to update a property in the dictionary.

Reply:
Ok, thanks for explaining! If it's not a json then we just might want to give it a different extension.